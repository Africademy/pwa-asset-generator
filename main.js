const pwa = require('./helpers/pwa');
const puppets = require('./puppets');
const flags = require('./helpers/flags');
const preLogger = require('./helpers/logger');

const generateImages = async (source, _output, _options, loggerFn) => {
  const logger = loggerFn || preLogger(generateImages.name, _options);

  if (!source) {
    throw Error('Please specify a URL or file path as a source');
  }

  const options = {
    ...flags.getDefaultOptions(),
    ..._options,
    ...flags.normalizeOnlyFlagPairs('splashOnly', 'iconOnly', _options, logger),
    ...flags.normalizeOnlyFlagPairs(
      'landscapeOnly',
      'portraitOnly',
      _options,
      logger,
    ),
  };

  const output = flags.normalizeOutput(_output);

  const savedImages = await puppets.generateImages(source, output, options);
  const manifestJsonContent = pwa.generateIconsContentForManifest(
    savedImages,
    options.manifest,
  );
  const htmlContent = pwa.generateHtmlForIndexPage(savedImages, options.index);

  if (!options.splashOnly) {
    if (options.manifest) {
      await pwa.addIconsToManifest(manifestJsonContent, options.manifest);
      logger.success(
        `Icons are saved to Web App Manifest file ${options.manifest}`,
      );
    } else if (!options.splashOnly) {
      logger.warn(
        'Web App Manifest file is not specified, printing out the content to console instead',
      );
      logger.success(
        'Below is the icons content for your manifest.json file. You can copy/paste it manually',
      );
      logger.raw(`\n${JSON.stringify(manifestJsonContent, null, 2)}\n\n`);
    }
  }

  if (options.index) {
    await pwa.addMetaTagsToIndexPage(htmlContent, options.index);
    logger.success(
      `iOS meta tags are saved to index html file ${options.index}`,
    );
  } else {
    logger.warn(
      'Index html file is not specified, printing out the content to console instead',
    );
    logger.success(
      'Below is the iOS meta tags content for your index.html file. You can copy/paste it manually',
    );
    logger.raw(`\n${htmlContent}\n`);
  }

  return { savedImages, htmlContent, manifestJsonContent };
};

module.exports = {
  generateImages,
};
