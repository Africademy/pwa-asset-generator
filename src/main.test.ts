import { generateImages } from './main';

const TEST_TIMEOUT_IN_MILLIS = 90000;

test('generates icons and splash screens when called via function', async () => {
  jest.setTimeout(TEST_TIMEOUT_IN_MILLIS);

  const result = await generateImages('./static/logo.png', './temp', {
    scrape: false,
    iconOnly: true,
    splashOnly: true,
    log: false,
    path: '%PUBLIC_URL%',
  });

  expect(result).toMatchSnapshot();
});
