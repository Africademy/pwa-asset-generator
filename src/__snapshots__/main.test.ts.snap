// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates icons and splash screens when called via function 1`] = `
Object {
  "htmlContent": "<link rel=\\"apple-touch-icon\\" sizes=\\"180x180\\" href=\\"%PUBLIC_URL%/temp/apple-icon-180.png\\">
<link rel=\\"apple-touch-icon\\" sizes=\\"167x167\\" href=\\"%PUBLIC_URL%/temp/apple-icon-167.png\\">
<link rel=\\"apple-touch-icon\\" sizes=\\"152x152\\" href=\\"%PUBLIC_URL%/temp/apple-icon-152.png\\">
<link rel=\\"apple-touch-icon\\" sizes=\\"120x120\\" href=\\"%PUBLIC_URL%/temp/apple-icon-120.png\\">

<meta name=\\"apple-mobile-web-app-capable\\" content=\\"yes\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-2048-2732.png\\"
    media=\\"(device-width: 1024px) and (device-height: 1366px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-2732-2048.png\\"
    media=\\"(device-width: 1024px) and (device-height: 1366px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-1668-2388.png\\"
    media=\\"(device-width: 834px) and (device-height: 1194px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-2388-1668.png\\"
    media=\\"(device-width: 834px) and (device-height: 1194px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-1668-2224.png\\"
    media=\\"(device-width: 834px) and (device-height: 1112px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-2224-1668.png\\"
    media=\\"(device-width: 834px) and (device-height: 1112px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-1536-2048.png\\"
    media=\\"(device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-2048-1536.png\\"
    media=\\"(device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-1242-2688.png\\"
    media=\\"(device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 3) and (orientation: portrait)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-2688-1242.png\\"
    media=\\"(device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 3) and (orientation: landscape)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-1125-2436.png\\"
    media=\\"(device-width: 375px) and (device-height: 812px) and (-webkit-device-pixel-ratio: 3) and (orientation: portrait)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-2436-1125.png\\"
    media=\\"(device-width: 375px) and (device-height: 812px) and (-webkit-device-pixel-ratio: 3) and (orientation: landscape)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-828-1792.png\\"
    media=\\"(device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-1792-828.png\\"
    media=\\"(device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-1242-2208.png\\"
    media=\\"(device-width: 414px) and (device-height: 736px) and (-webkit-device-pixel-ratio: 3) and (orientation: portrait)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-2208-1242.png\\"
    media=\\"(device-width: 414px) and (device-height: 736px) and (-webkit-device-pixel-ratio: 3) and (orientation: landscape)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-750-1334.png\\"
    media=\\"(device-width: 375px) and (device-height: 667px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-1334-750.png\\"
    media=\\"(device-width: 375px) and (device-height: 667px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-640-1136.png\\"
    media=\\"(device-width: 320px) and (device-height: 568px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)\\">
<link rel=\\"apple-touch-startup-image\\"
    href=\\"%PUBLIC_URL%/temp/apple-splash-1136-640.png\\"
    media=\\"(device-width: 320px) and (device-height: 568px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)\\">
",
  "manifestJsonContent": Array [
    Object {
      "sizes": "192x192",
      "src": "temp/manifest-icon-192.png",
      "type": "image/png",
    },
    Object {
      "sizes": "512x512",
      "src": "temp/manifest-icon-512.png",
      "type": "image/png",
    },
  ],
  "savedImages": Array [
    Object {
      "height": 2732,
      "name": "apple-splash-2048-2732",
      "orientation": "portrait",
      "path": "temp/apple-splash-2048-2732.png",
      "scaleFactor": 2,
      "width": 2048,
    },
    Object {
      "height": 2048,
      "name": "apple-splash-2732-2048",
      "orientation": "landscape",
      "path": "temp/apple-splash-2732-2048.png",
      "scaleFactor": 2,
      "width": 2732,
    },
    Object {
      "height": 2388,
      "name": "apple-splash-1668-2388",
      "orientation": "portrait",
      "path": "temp/apple-splash-1668-2388.png",
      "scaleFactor": 2,
      "width": 1668,
    },
    Object {
      "height": 1668,
      "name": "apple-splash-2388-1668",
      "orientation": "landscape",
      "path": "temp/apple-splash-2388-1668.png",
      "scaleFactor": 2,
      "width": 2388,
    },
    Object {
      "height": 2224,
      "name": "apple-splash-1668-2224",
      "orientation": "portrait",
      "path": "temp/apple-splash-1668-2224.png",
      "scaleFactor": 2,
      "width": 1668,
    },
    Object {
      "height": 1668,
      "name": "apple-splash-2224-1668",
      "orientation": "landscape",
      "path": "temp/apple-splash-2224-1668.png",
      "scaleFactor": 2,
      "width": 2224,
    },
    Object {
      "height": 2048,
      "name": "apple-splash-1536-2048",
      "orientation": "portrait",
      "path": "temp/apple-splash-1536-2048.png",
      "scaleFactor": 2,
      "width": 1536,
    },
    Object {
      "height": 1536,
      "name": "apple-splash-2048-1536",
      "orientation": "landscape",
      "path": "temp/apple-splash-2048-1536.png",
      "scaleFactor": 2,
      "width": 2048,
    },
    Object {
      "height": 2688,
      "name": "apple-splash-1242-2688",
      "orientation": "portrait",
      "path": "temp/apple-splash-1242-2688.png",
      "scaleFactor": 3,
      "width": 1242,
    },
    Object {
      "height": 1242,
      "name": "apple-splash-2688-1242",
      "orientation": "landscape",
      "path": "temp/apple-splash-2688-1242.png",
      "scaleFactor": 3,
      "width": 2688,
    },
    Object {
      "height": 2436,
      "name": "apple-splash-1125-2436",
      "orientation": "portrait",
      "path": "temp/apple-splash-1125-2436.png",
      "scaleFactor": 3,
      "width": 1125,
    },
    Object {
      "height": 1125,
      "name": "apple-splash-2436-1125",
      "orientation": "landscape",
      "path": "temp/apple-splash-2436-1125.png",
      "scaleFactor": 3,
      "width": 2436,
    },
    Object {
      "height": 1792,
      "name": "apple-splash-828-1792",
      "orientation": "portrait",
      "path": "temp/apple-splash-828-1792.png",
      "scaleFactor": 2,
      "width": 828,
    },
    Object {
      "height": 828,
      "name": "apple-splash-1792-828",
      "orientation": "landscape",
      "path": "temp/apple-splash-1792-828.png",
      "scaleFactor": 2,
      "width": 1792,
    },
    Object {
      "height": 2208,
      "name": "apple-splash-1242-2208",
      "orientation": "portrait",
      "path": "temp/apple-splash-1242-2208.png",
      "scaleFactor": 3,
      "width": 1242,
    },
    Object {
      "height": 1242,
      "name": "apple-splash-2208-1242",
      "orientation": "landscape",
      "path": "temp/apple-splash-2208-1242.png",
      "scaleFactor": 3,
      "width": 2208,
    },
    Object {
      "height": 1334,
      "name": "apple-splash-750-1334",
      "orientation": "portrait",
      "path": "temp/apple-splash-750-1334.png",
      "scaleFactor": 2,
      "width": 750,
    },
    Object {
      "height": 750,
      "name": "apple-splash-1334-750",
      "orientation": "landscape",
      "path": "temp/apple-splash-1334-750.png",
      "scaleFactor": 2,
      "width": 1334,
    },
    Object {
      "height": 1136,
      "name": "apple-splash-640-1136",
      "orientation": "portrait",
      "path": "temp/apple-splash-640-1136.png",
      "scaleFactor": 2,
      "width": 640,
    },
    Object {
      "height": 640,
      "name": "apple-splash-1136-640",
      "orientation": "landscape",
      "path": "temp/apple-splash-1136-640.png",
      "scaleFactor": 2,
      "width": 1136,
    },
    Object {
      "height": 180,
      "name": "apple-icon-180",
      "orientation": null,
      "path": "temp/apple-icon-180.png",
      "scaleFactor": undefined,
      "width": 180,
    },
    Object {
      "height": 167,
      "name": "apple-icon-167",
      "orientation": null,
      "path": "temp/apple-icon-167.png",
      "scaleFactor": undefined,
      "width": 167,
    },
    Object {
      "height": 152,
      "name": "apple-icon-152",
      "orientation": null,
      "path": "temp/apple-icon-152.png",
      "scaleFactor": undefined,
      "width": 152,
    },
    Object {
      "height": 120,
      "name": "apple-icon-120",
      "orientation": null,
      "path": "temp/apple-icon-120.png",
      "scaleFactor": undefined,
      "width": 120,
    },
    Object {
      "height": 192,
      "name": "manifest-icon-192",
      "orientation": null,
      "path": "temp/manifest-icon-192.png",
      "scaleFactor": undefined,
      "width": 192,
    },
    Object {
      "height": 512,
      "name": "manifest-icon-512",
      "orientation": null,
      "path": "temp/manifest-icon-512.png",
      "scaleFactor": undefined,
      "width": 512,
    },
  ],
}
`;
